### PR TITLE
[prim] Update fifo behavior during reset

### PR DESCRIPTION
**Pull request by tjaychen**
Address #4260

Signed-off-by: Timothy Chen <timothytim@google.com>

### CODE REVIEW

**Comment by tjaychen** [comment]
hmmm.... i'm guessing we have some assumptions about being able to immediately write fifos coming out of reset.
I'll have a look.

**Comment by tjaychen** [comment]
I took a look at the error, and the problem basically boils down to `wready_o` being an overloaded function for `full`.
During reset, `wready_o` is now 0 due to this change.  This makes much of the logic think the FIFO is "full", and this trickles a lot of downstream things in `flash_ctrl` and the entropy blocks, which is where we are getting the errors.

To resolve this, I think we need to once again de-couple the concept of `wready_o` and `full`. The former should be 0 during reset, while the latter remains more of a status indication of the FIFO.

Before I do that and change all these modules though, wanted to get some feedback from people to see what you guys all think. As mentioned in the issue, the "reset out of sync" issue is not likely going to be present on OpenTitan, although I can't say for sure it does not exist. I imagine it could happen as long as resets are released on differently gated clocks (even if they are synchronous).  Then the reset release could be separate by at least one cycle (more so for async domains).  So if a host immediately tries to issue a transaction, it COULD be a problem.

I feel like we should just update it, and I can go through and make all the required changes (sync now, async later).

**Comment by msfschaffner** [comment]
Yes, I also think separating flow control (ready/valid) from status indication (full/empty) is cleaner.

**Comment by mwbranstad** [comment]
I suppose making wready_o be zero during reset is accurate, and the status should always be accurate (empty in this case). Making any change to improve reset behavior is good. However, it seems very precarious to design a system or chip that can even allow timing this close to reset de-asserting. There is typically some boot sequencer that would provide signals to move full operation of some logic away from reset. I think OT has this. The rule of thumb that i am aware of is that guarantying logic to coming out of reset on the same exact clock cycle is very difficult to enforce.

**Comment by tjaychen** [comment]
definitely agree with you Mark, and we do have that, generally speaking.
Especially since at the moment our only host is the processor and
potentially the ROM checker.

**Comment by tjaychen** [comment]
@mwbranstad are you okay with this overall change then? I can go ahead and make the needed changes in entropy_src, csrng and others.  And then push for you to review.

**Comment by mwbranstad** [comment]
@tjaychen I am now confused. I thought the prim_fifo_sync change would be contained within the FIFO block, and that it would be transparent to the parent modules. Can you add a few words about why this is not the case, and what changes you think are needed at the instantiation level? If there are some, I think I would prefer to make the changes.

**Comment by tjaychen** [comment]
> @tjaychen I am now confused. I thought the prim_fifo_sync change would be contained within the FIFO block, and that it would be transparent to the parent modules. Can you add a few words about why this is not the case, and what changes you think are needed at the instantiation level? If there are some, I think I would prefer to make the changes.

ah yes, let me elaborate.
The changes should be within the `prim_fifo_sync`. However, the problem is that right now, multiple modules (`flash_ctrl`, `csrng` for example) use `~wready_o` to indicate `full`.  This causes a problem because during reset full is now `1`.

This in flash causes the `full` interrupt to erroneously assert even though there is nothing there.  In `csrng`, this triggers the fifo error interrupt because a fifo is now both `full` and `empty` at the same time.  The change I've added on the fifo is that during reset, both `rvalid_o` and `wready_o` are 0's.  I'm qualifying the `rvalid_o` case also because i've seen in different situations where cases like this cause transaction duplication (as well as transaction drops). 

The changes are really minor, I don't mind doing them and just pushing in review for you to have a look.  If you're not comfortable with the change, I am happy to roll it back.  I do need to do something though otherwise CI won't work :)

**Comment by mwbranstad** [comment]
Ok, sounds good. Make any changes and I will look at them.

**Comment by mwbranstad** [comment]
Also, could i get pr #4584 merged before you make the changes? May not be affected, but would like to have the rtl in sync.

**Comment by tjaychen** [comment]
ah dont worry i won't merge this one before that.

**Comment by tjaychen** [comment]
@mwbranstad please take a look. It seems like in some cases the `wready` is not really used outside of error indication, so it would be good if you can double check i didn't do anything wrong.

**Comment by mwbranstad** [comment]
@tjaychen thanks for making these sample changes...but...I don't like this strategy in general. CSRNG uses this FIFO in spades, and uses the wready_o pin much more functionally than EDN does. I would prefer we leave the FIFO io pins alone, and we patch the reset condition on a per use basis. For example, the "bug" in EDN, CSRNG, ES is that the fifo_err signal is not gated with the enable signal, which insures that the reset condition is unaffected by the timing of reset.  I think we should consider this type of fix for the other cases that are failing. I am just afraid that this change will not only be a pain to make everywhere, but affect the debug of these designs, which has not begun in earnest yet. What do you think of this approach?

**Comment by tjaychen** [comment]
ah yes that's fine with me too.
There's no reason this HAS to be fixed a specific way, and what you're
proposing certainly seems to disturb the existing logic a lot less.

I can back-out these changes and switch to enable, could you let me know
what signal I should use in each case..? Or if you prefer to do a
separate PR for that first, I can then rebase on top of yours for this one.

**Comment by mwbranstad** [comment]
Go ahead back out your changes. I will make a separate PR with the changes for the 3 entropy blocks.

**Comment by tjaychen** [comment]
sounds good.
I'll wait for your PR before merging.

**Comment by tjaychen** [comment]
@mwbranstad i had a look at the latest failures after rebasing (sorry it took me awhile), it looks like the interrupts are okay now, but the err_code registers are picking up the error value right out of reset.

Do you want to add an `enable` qualification to the err_code `de` as well? Similar to what you did for the interrupt path?

**Comment by mwbranstad** [comment]
Easy enough to do - will add @tjaychen

**Comment by mwbranstad** [comment]
@tjaychen I have PRs #4824, #4843, and #4844 that add the enable to the ERR_CODE register update.

**Comment by tjaychen** [comment]
thanks! let me pull them into my area and see if all the related tests pass.

**Comment by tjaychen** [comment]
@mwbranstad all fixes look good!
I've merged two of yours, so let me know on the last one.  
Once that is in, I will rebase and we can work on merging this one.

**Comment by mwbranstad** [comment]
@tjaychen good to hear!

**Comment by tjaychen** [comment]
could someone give this a quick review?

File: hw/ip/entropy_src/doc/_index.md
Comment (martin-lueker) [inline hw/ip/entropy_src/doc/_index.md:137]: Is the vendor-specific test still planned?

File: hw/ip/entropy_src/rtl/entropy_src_core.sv
Comment (martin-lueker) [inline hw/ip/entropy_src/rtl/entropy_src_core.sv:92]: Super minor nits, but the alignment here could be improved.

**Review by martin-lueker — APPROVED** [review]
Comment on the figure: This seems to have carried over from the original, but the register now called fw_ov_rd_data.d, it's not super clear what is sourcing it.   Is there any way to put a dot at the output of the fifo to highlight that that signal is going to three places?

Other than that, I think the only issue is formatting and alignment.  It would be worthwhile just making a scan for all of those.
Other than that LGTM.

File: hw/ip/entropy_src/doc/_index.md
Comment (mwbranstad) [reply hw/ip/entropy_src/doc/_index.md:137]: Yes, diagram shows this. Added text to explain the interface in the doc section.

File: hw/ip/entropy_src/rtl/entropy_src_core.sv
Comment (mwbranstad) [reply hw/ip/entropy_src/rtl/entropy_src_core.sv:92]: fixed up to align that chunk of declarations.

File: hw/ip/entropy_src/doc/_index.md
Comment (mwbranstad) [reply hw/ip/entropy_src/doc/_index.md:137]: Also, some dots were added to show more clarity on bus crossings.

File: hw/ip/prim/rtl/prim_fifo_sync.sv
Comment (msfschaffner) [inline hw/ip/prim/rtl/prim_fifo_sync.sv:30]: if we're adding a dedicated a full flag here, would it make sense to output the empty flag as well for consistency? I believe the signal is already available internally...

File: hw/ip/prim/rtl/prim_fifo_sync.sv
Comment (msfschaffner) [inline hw/ip/prim/rtl/prim_fifo_sync.sv:68]: nit: is there a specific reason this uses an inverter? i.e., would the following work as well?
```systemverilog
      end else begin
        under_rst <= 1'b0;
      end
```

**Review by msfschaffner — APPROVED** [review]
FIFO changes LGTM with two small comments.

File: hw/ip/prim/rtl/prim_fifo_sync.sv
Comment (tjaychen) [reply hw/ip/prim/rtl/prim_fifo_sync.sv:68]: yeah i can switch to that.

File: hw/ip/prim/rtl/prim_fifo_sync.sv
Comment (tjaychen) [reply hw/ip/prim/rtl/prim_fifo_sync.sv:30]: yeah, i just didn't add it because...no one was using it heh. 
i can add that as a separate PR

**Pull request merged**
